test/mocha/10-api.js

```diff
@@ -1,3 +1,745 @@
+        const challenge = authenticator.generate(secret);
+
+        // authenticate with the token
+        let res;
+        try {
+          res = await httpClient.post(`${authenticateURL}`, {
+            agent,
+            json: {
+              account: accountId,
+              type,
+              challenge,
+            }
+          });
+        } catch(e) {
+          err = e;
+        }
+        assertNoError(err);
+        should.exist(res);
+        should.exist(res.data);
+        res.data.should.be.an('object');
+        res.data.authenticated.should.equal(true);
+        res.data.authenticatedMethods.should.be.an('array');
+        res.data.authenticatedMethods[0].should.equal(
+          'token-client-registration'
+        );
+        res.status.should.equal(200);
+        should.exist(res.headers.get('set-cookie'));
+        // check if token client  is registered.
+        let res2;
+        let err2;
+        const splitCookieHeaders =
+          setCookie.splitCookiesString(res.headers.get('set-cookie'));
+        const cookies = setCookie.parse(splitCookieHeaders);
+        try {
+          res2 = await httpClient.get(
+            `${registrationURL}?email=alpha@example.com`, {
+              agent, headers: {
+                Cookie: `${cookies[0].name}=${cookies[0].value}`
+              }
+            });
+        } catch(e) {
+          err = e;
+        }
+        assertNoError(err2);
+        should.exist(res2);
+        res2.data.registered.should.equal(true);
+      });
+    it('should return `false` if client is not registered.',
+      async function() {
+        // call the endpoint with an unregistered token client.
+        let res;
+        let err;
+        const {agent} = brHttpsAgent;
+        try {
+          res = await httpClient.get(
+            `${registrationURL}?email=beta@example.com`, {
+              agent
+            });
+        } catch(e) {
+          err = e;
+        }
+        assertNoError(err);
+        should.exist(res);
+        res.data.registered.should.equal(false);
+      });
+  });
+  describe('get /routes.login', () => {
+    beforeEach(async function setup() {
+      await helpers.prepareDatabase(mockData);
+      actors = await helpers.getActors(mockData);
+      accounts = mockData.accounts;
+    });
+    afterEach(async function() {
+      // replace stub with an empty stub.
+      passportStub.callsFake((req, res, next) => next());
+    });
+    it('should successfully login with multifactor authentication',
+      async function() {
+        const type = 'totp';
+        const {agent} = brHttpsAgent;
+        stubPassportStub('alpha@example.com');
+        const accountId = accounts['alpha@example.com'].account.id;
+        const actor = await brAccount.getCapabilities({id: accountId});
+
+        await brAuthnToken.setAuthenticationRequirements({
+          account: accountId,
+          actor,
+          requiredAuthenticationMethods: ['token-client-registration'],
```

Comment:
I believe in order to be multi-factor you need 2 authentication methods such as `password` and `otop` or `email-link` and `totp` etc. @dlongley am I wrong here?

Multifactor authentication:

https://www.nist.gov/itl/applied-cybersecurity/tig/back-basics-multi-factor-authentication

https://en.wikipedia.org/wiki/Multi-factor_authentication

@JSAssassin so I think you need 2 required authentication methods here, not one.

I might be wrong here as `token-client-registration` could be two auth methods in one.

Reply:
The API accepts any number of arguments. While best security practices are to have more than one, it would be awkward to force that here and it wouldn't cover all the use cases (or would require a separate API where only one is used).

Reply:
yes, but the test is called `should successfully login with multifactor authentication`, but it appears to only use one form of authentication. Is having one one `requiredAuthenticationMethods` and one authentication somehow multifactor here?